_posts/2021-09-29-signed-immutables-bug.md

```diff
@@ -0,0 +1,92 @@
+---
+layout: post
+published: true
+title: 'Signed Immutables Bug'
+date: '2021-09-29'
+author: Solidity Team
+category: Security Alerts
+---
+
+On September 28th, 2021, the Solidity team discovered that
+for immutable variables of a signed integer type shorter than 256 bits,
+sign extension (cleanup) of its value is not always properly performed.
+
+Inline assembly is required to access the value in its unclean state.
+The bug is present since the introduction of the
+`immutable` feature in Solidity 0.6.5 and is fixed in 0.8.9.
+
+We assigned the bug a severity of "very low".
+
+## Technical Details
+
+When immutable variables are assigned in Solidity during the construction phase, the value
+is stored in memory until the creation concludes. For types shorter than 32 bytes,
+these values were stored left-aligned within the 32 byte word and when the values were
+read again, they were shifted back accordingly, but no cleanup was performed on the values.
+In particular, the signextend operation was not performed on signed values.
+
+As an example, if you store ``-2`` in an ``int8``, it is stored in two's complement encoding
+as a single-byte value of ``0xfe``. Since the words in the EVM use 32 bytes, the value
+needs to be properly sign-extended before an operation like division is performed,
+otherwise the value would be interpreted as a positive value.
+
+Sign-extension means setting the higher order bits to the same value as the
+sign bit, which is the most significant bit of the actual type.
+
+The sign-extension of ``0xfe`` is `0xffffffff..fffffffe` (full 32 bytes),
```

Comment:
```suggestion
The sign-extension of the value ``0xfe`` of type ``int8`` is `0xffffffff..fffffffe` (full 32 bytes),
```
Somehow want to mention that this is about the type `int8`.

Reply:
Or maybe something more detailed, like:

The value `-2` represented as a twos complement in 16 bits (`int8`) is `0xfe`, however, in EVM, it is sign extended to `0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe`), which also represents the value `-2` when interpreted as `int25`. Note that `0xfe` when interpreted as `int256` is `+254`.